heft-plugins/heft-jest-plugin/src/JestPlugin.ts

```diff
@@ -159,84 +183,215 @@ export default class JestPlugin implements IHeftTaskPlugin<IJestPluginOptions> {
-    terminal.writeLine(`Using Jest version ${getVersion()}`);
 
-    // Write the jest data file used by the BuildTransformer
-    await this._setupJestAsync(taskSession, heftConfiguration, options);
+    const pendingTestRuns: Set<IPendingTestRun> = this._pendingTestRuns;
+
+    if (!this._jestPromise) {
+      // Monkey-patch Jest's watch mode so that we can orchestrate it.
+      const jestCoreDir: string = path.dirname(require.resolve('@jest/core'));
+
+      const wrappedStdOut: TerminalWritableStream = new TerminalWritableStream(terminal);
+
+      // Shim watch so that we can intercept the output stream
+      const watchModulePath: string = path.resolve(jestCoreDir, 'watch.js');
+      const watchModule: {
+        default: IJestWatch;
+      } = require(watchModulePath);
+      const { default: originalWatch } = watchModule;
+
+      const watch: IJestWatch = (
+        initialGlobalConfig: unknown,
+        contexts: unknown,
+        outputStream: NodeJS.WriteStream,
+        hasteMapInstances: unknown,
+        stdin: NodeJS.ReadStream | undefined,
+        hooks: unknown,
+        filter: unknown
+      ) => {
+        return originalWatch(
+          initialGlobalConfig,
+          contexts,
+          wrappedStdOut as unknown as NodeJS.WriteStream,
+          hasteMapInstances,
+          stdin,
+          hooks,
+          filter
+        );
+      };
+      Object.defineProperty(watchModule, 'default', {
+        get(): typeof originalWatch {
+          return watch;
+        }
+      });
+
+      // Shim runJest so that we can defer test execution
+      const runJestModulePath: string = path.resolve(jestCoreDir, 'runJest.js');
+      const runJestModule: {
+        default: (params: IRunJestParams) => Promise<void>;
+      } = require(runJestModulePath);
+      const { default: originalRunJest } = runJestModule;
+      const runJest: typeof originalRunJest = (params: IRunJestParams): Promise<void> => {
+        if (!this._executing) {
+          setTimeout(requestRun, 100);
+        }
+
+        return new Promise((resolve: () => void, reject: (err: Error) => void) => {
+          pendingTestRuns.add(async (): Promise<AggregatedResult | undefined> => {
+            let result: AggregatedResult | undefined;
+            const { onComplete } = params;
+            try {
+              await originalRunJest({
+                ...params,
+                onComplete: (testResults: AggregatedResult) => {
+                  result = testResults;
+                  onComplete(testResults);
+                }
+              });
+              resolve();
+            } catch (err) {
+              reject(err);
+              throw err;
+            }
+
+            return result;
+          });
+        });
+      };
+
+      Object.defineProperty(runJestModule, 'default', {
+        get(): typeof originalRunJest {
+          return runJest;
+        }
+      });
+
+      const { getVersion, runCLI } = await import(`@jest/core`);
```

Comment:
Dedupe

Reply:
Postponing since the Jest integration may get redone later to avoid global state pollution.